File: internal/methods/lagrange_interpolation_method.go
package methods

import (
	"errors"

	"github.com/jictyvoo/tec217_computing-methods/internal/models"
)

type LagrangeInterpolationMethod[T models.Numeric] struct {
	commonFuncZeroState[T]
}

func (mtd *LagrangeInterpolationMethod[T]) Run(x, y []T, xx T) (result T, err error) {
	if len(x) != len(y) {
		err = errors.New("x and y must have the same length")
		return
	}

	sampleSize := len(x)

	for ctrlIndex := 0; ctrlIndex < sampleSize; ctrlIndex++ {
		term := y[ctrlIndex]

		// Start to perform the prod calculation
		for index := 0; index < sampleSize; index++ {
			if ctrlIndex != index {
				term *= (xx - x[index]) / (x[ctrlIndex] - x[index])
			}
		}
		result += term
		mtd.registerInteraction(x, uint32(ctrlIndex), 0, result, xx)
	}

	return
}

File: scripts/octave/methods/lagrangeInterpolationMethod.m
function [result, err] = lagrangeInterpolationMethod(x, y, xx)
    err = ""; % Initialize err to an empty string
    sampleSize = length(x);

    % Check for consistent input
    if sampleSize != length(y)
        err = "x and y must have the same length";
        result = NaN;
        return;
    endif

    result = 0;
    for ctrlIndex = 1:sampleSize
        term = y(ctrlIndex);

        % Start to perform the prod calculation
        for index = 1:sampleSize
            if ctrlIndex ~= index
                term *= (xx - x(index)) / (x(ctrlIndex) - x(index));
            end
        end
        result += term;
    end

endfunction
